### PR TITLE
add user groups to outputs

### PR DESCRIPTION
**Pull request by rtpascual**
## Changes

Add user groups and their precedence to `amplify_outputs.json`.

Example:
```ts
export const auth = defineAuth({
  loginWith: {
    email: true,
  },
  groups: ['Administrators', 'Editors']
});
```

```json
{
  "auth": {
    ...
    "groups": [
      {
        "Administrators": {
          "precedence": 0
        }
      },
      {
        "Editors": {
          "precedence": 1
        }
      }
    ]
  },
  ...
}
```

**Corresponding docs PR, if applicable:**

## Validation

E2E and unit tests

## Checklist

- [ ] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

**Review by sobolk — DISMISSED** [review]

File: packages/client-config/src/client-config-contributor/client_config_contributor_factory.ts [hunk 1]
```diff
@@ -58,12 +68,12 @@ export class ClientConfigContributorFactory {
         new Custom1_1(),
       ],
 
-      // Legacy config is derived from V1.2 (latest) of unified default config
+      // Legacy config is derived from V1.3 (latest) of unified default config
       [ClientConfigVersionOption.V0]: [
         new Auth1_1(),
```
Comment (Amplifiyer) [inline packages/client-config/src/client-config-contributor/client_config_contributor_factory.ts:73]: Do we have `groups` in Gen1 config?

File: packages/client-config/src/client-config-contributor/client_config_contributor_v1.ts [hunk 2]
```diff
@@ -58,9 +72,181 @@ export class VersionContributorV1 implements ClientConfigContributor {
...
+  };
+}
+
+/**
+ * Translator for the Auth portion of ClientConfig in V1.2
+ */
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export class AuthClientConfigContributorV1_1
+  implements ClientConfigContributor
+{
```
Comment (Amplifiyer) [inline packages/client-config/src/client-config-contributor/client_config_contributor_v1.ts:249]: Do we need to create a new contributor here? If the new field is optional, won't updating the existing one would work for generating both new and old configs?

File: packages/client-config/src/client-config-schema/client_config_v1.3.ts [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+  custom?: {
+    [k: string]: unknown;
+  };
+}
+/**
+ * This interface was referenced by `undefined`'s JSON-Schema definition
+ * via the `patternProperty` ".*".
+ */
+export interface AmplifyUserGroupConfig {
+  precedence?: number;
```
Comment (Amplifiyer) [inline packages/client-config/src/client-config-schema/client_config_v1.3.ts:248]: should this be required?

File: packages/auth-construct/src/construct.test.ts [hunk 4]
```diff
@@ -1480,6 +1481,34 @@ void describe('Auth construct', () => {
...
       assert.equal(outputs['socialProviders']['Value'], `["GOOGLE"]`);
     });
+    void it('can override group precedence and correctly updates stored output', () => {
+      const app = new App();
+      const stack = new Stack(app);
+      const auth = new AmplifyAuth(stack, 'test', {
+        loginWith: { email: true },
+        groups: ['admins', 'managers'],
+      });
+      auth.resources.groups['admins'].cfnUserGroup.precedence = 2;
```
Comment (Amplifiyer) [inline packages/auth-construct/src/construct.test.ts:1491]: What happens if this was overridden to `1`, i.e. conflict with the other group's precedence?

File: packages/client-config/src/client-config-contributor/client_config_contributor_factory.ts (hunk 1 above)
Comment (rtpascual) [reply packages/client-config/src/client-config-contributor/client_config_contributor_factory.ts:73]: I believe for the Gen1 config it is `userPoolGroups`

File: packages/client-config/src/client-config-contributor/client_config_contributor_v1.ts (hunk 2 above)
Comment (rtpascual) [reply packages/client-config/src/client-config-contributor/client_config_contributor_v1.ts:249]: I followed the same pattern when we added storage access paths to outputs, the new field `paths` is optional but we still created a new contributor. This also minimizes the changes of us breaking existing versions.

File: packages/auth-construct/src/construct.test.ts (hunk 4 above)
Comment (rtpascual) [reply packages/auth-construct/src/construct.test.ts:1491]: For us this is possible and Cognito allows it, only issue is in the frontend when they get credentials, the `preferred_role` is not set if the role ARN is different. See https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-user-groups.html#assigning-precedence-values-to-groups.

File: packages/client-config/src/client-config-contributor/client_config_contributor_factory.ts (hunk 1 above)
Comment (Amplifiyer) [reply packages/client-config/src/client-config-contributor/client_config_contributor_factory.ts:73]: Then should we add that conversion for legacy config as well?

File: packages/client-config/src/client-config-contributor/client_config_contributor_v1.ts (hunk 2 above)
Comment (Amplifiyer) [reply packages/client-config/src/client-config-contributor/client_config_contributor_v1.ts:249]: But we are duplicating quite a bit of code. Can we assert from tests that we won't break existing versions?

File: packages/client-config/src/client-config-contributor/client_config_contributor_factory.ts (hunk 1 above)
Comment (rtpascual) [reply packages/client-config/src/client-config-contributor/client_config_contributor_factory.ts:73]: Sorry looked at the wrong file, we don't have groups in Gen1 config.

**Review by sobolk — APPROVED** [review]

**Review by Amplifiyer — APPROVED** [review]

**Review by sobolk — APPROVED** [review]

**Review by Amplifiyer — APPROVED** [review]

**Review by Amplifiyer — APPROVED** [review]

**Pull request merged**